### DISCUSSION TITLE
How do we implement `Ratio` and `Complex`?

### DISCUSSION THREAD

**Discussion by stylewarning** [coalton-lang/coalton#155] [Ideas and Proposals]
There is difficulty in implementing these as native Lisp objects, because:

- Problem 1: reduction of `#C(k 0)` to `k`; or `k/1` to `k`

- Problem 2: Lisp doesn't allow arbitrary field elements as components (`Complex :a` cannot use Lisp's complex numbers)

How do we implement `Complex Single-Float` as specialized on `(COMPLEX SINGLE-FLOAT)`, but also allow `Num :a => Complex :a` to exist?

**Comment by stylewarning** [comment]
See also [`lisp-type`](https://github.com/coalton-lang/coalton/blob/main/src/codegen/lisp-types.lisp#L7)

**Comment by bitmappergit** [comment]
I don't think Ratio is an issue once #51 is solved due to the reduction of those to integers being the first thing that happens, which would then have something like ((Num :a) => :a) for their type, and ratios must have an instance of Num to implement Fractional.

**Comment by eliaslfox** [comment]
I think @stylewarning is referring to the fact that SBCL will convert `#C(k 0)` into a real type at runtime.

**Comment by stylewarning** [comment]
Yes:

```lisp
CL-USER> 4/3
4/3
CL-USER> (type-of *)
RATIO
CL-USER> 4/2
2
CL-USER> (type-of *)
(INTEGER 0 4611686018427387903)
CL-USER> #C(2 1)
#C(2 1)
CL-USER> (type-of *)
(COMPLEX (INTEGER 1 2))
CL-USER> #C(2 0)
2
CL-USER> (type-of *)
(INTEGER 0 4611686018427387903)
```

**Comment by bitmappergit** [comment]
Ah, sorry, I misunderstood the issue.

It might just be best to represent these as a datatype like in Haskell, and handle this in the codegen?
